### PR TITLE
use `--reporter=failures-only` for CI tests that run from `dev/bots/test.dart`

### LINKED ISSUE #149611: [flutter_tools] Consider using `--reporter=failures-only` in flutter tool CI tests that use package:test
See title. This reporter (recently made available by https://github.com/flutter/flutter/pull/148739) might be useful to reducing noise in logs from CI tests.

### PR DESCRIPTION
**Pull request by andrewkolos**
Resolves https://github.com/flutter/flutter/issues/149611

~~Resolvs #27845 (since this affects Framework as well)~~ Removing this as I don't believe this change would affect integration tests.

(also contains some light refactors to some `expect` calls)

TODO 
- [x] Consider doing this for `flutter test` calls as well.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by andrewkolos** [comment]
I tested this out by intentionally breaking some tests (in commit [experiment: intentionally introduce some test failures to see how CI …](https://github.com/flutter/flutter/pull/150874/commits/e6e9c894456710737fcf82a2a5764c5dff871ee2)[experiment: intentionally introduce some test failures to see how CI …](https://github.com/flutter/flutter/pull/150874/commits/e6e9c894456710737fcf82a2a5764c5dff871ee2)).

The result seems...really good! Check out this failure:

https://github.com/flutter/flutter/pull/150874/checks?check_run_id=26728828337

You can see the failed test output without having to navigate to LUCI!

Compare this to the status-quo, a failing check from an open tools PR: 

https://github.com/flutter/flutter/pull/150850/checks?check_run_id=26712984904

File: dev/bots/utils.dart [hunk 1]
```diff
@@ -341,6 +341,7 @@ Future<void> runDartTest(String workingDirectory, {
   bool ensurePrecompiledTool = true,
   bool shuffleTests = true,
   bool collectMetrics = false,
+  String? reporter = 'expanded', // TODO-dontmerge-should we just always use --failures-only?
```
Comment (andrewkolos) [inline dev/bots/utils.dart:344]: Is there a pragmatic reason to not use `failures-only`? Should we just remove this parameter and always use `failures-only`?

Comment (andrewkolos) [reply dev/bots/utils.dart]: I'm going for it. Added commit [always use --reporter=failures-only](https://github.com/flutter/flutter/pull/150874/commits/2eb30cf46fbed66a8d56342a161fd2980fd8af52)[always use --reporter=failures-only](https://github.com/flutter/flutter/pull/150874/commits/2eb30cf46fbed66a8d56342a161fd2980fd8af52)

**Comment by andrewkolos** [comment]
Added @goderbauer has a reviewer since this would affect Framework contributors.

**Comment by goderbauer** [comment]
The new shortened test output looks very nice! I do have one question: Sometimes we've been adding extra logging to a test to track down some flakiness. For those cases it has always been helpful to compare the logs of a passing CI run with a failing CI run. Would that still be possible with this mode - or are all the runs for passing runs now thrown away?

**Comment by andrewkolos** [comment]
> The new shortened test output looks very nice! I do have one question: Sometimes we've been adding extra logging to a test to track down some flakiness. For those cases it has always been helpful to compare the logs of a passing CI run with a failing CI run. Would that still be possible with this mode - or are all the runs for passing runs now thrown away?

Great question. I don't think `--reporter=failures-only` intercepts any `print` calls, so I think that would still work?

Consider

```diff
// packages/flutter_tools/test/integration.shard/hot_reload_test.dart
  testWithoutContext('hot restart works without error', () async {
+   print('hello @goderbauer');
    await flutter.run();
    await flutter.hotRestart();
  });
```

```
andrewkolos-macbookpro:flutter_tools andrewkolos$ flutter test test/integration.shard/hot_reload_test.dart --reporter=failures-only
+3: hot restart works without error
hello @goderbauer
+6: All tests passed!
```

**Comment by christopherfujino** [comment]
Another drawback is that if a test library that is running single-threaded and one of them hangs, then seeing the last test that started in the logs is helpful.

**Comment by andrewkolos** [comment]
> Another drawback is that if a test library that is running single-threaded and one of them hangs, then seeing the last test that started in the logs is helpful.

True. Still, I wonder how often this would be a problem. Such a situation would mean that the test ended up busy-waiting (never deferring to the event loop). I'm still willing to go through with this change.

**Comment by andrewkolos** [comment]
@christopherfujino @goderbauer friendly ping

**Review by christopherfujino — APPROVED** [review]
LGTM if goderbauer is happy with this.

**Review by goderbauer — APPROVED** [review]
LGTM

Let's give this a try and keep an eye on whether this makes debugging test failures any harder. If it does, we can always roll this back.

**Comment by andrewkolos** [comment]
Closing to focus on tool crashes and flakes. If someone wants to pick this up, the follow-up would simply to be make an announcement (probably on discord) about this change and then merge it.

**Pull request closed without merging**